app/src/main/java/com/epfl/drawyourpath/utils/Utils.kt

```diff
@@ -392,4 +394,118 @@ object Utils {
+        return Path(reducedPoints.toList())
+    }
+
+    /**
+     * Reduces the number of points in a section.
+     * @param section The section we want to reduce
+     * @param epsilon The min distance to keep the point on the section
+     * @return A similar segment with the least useful points removed.
+     */
+    fun reduceSection(section: List<LatLng>, epsilon: Float): List<LatLng> {
+        val reducedPointList = mutableListOf<LatLng>()
+        if (section.size <= 2) {
+            reducedPointList.addAll(section)
+            return reducedPointList.toList()
+        }
+
+        val distances = distancesToSegment(section.subList(1, section.size - 1), section.first(), section.last())
+        val distMax = distances.max()
+        val indexMax = distances.indexOf(distMax) + 1
+        // Check if finished, otherwise we need to solve recursively
+        if (distMax <= epsilon) {
+            reducedPointList.add(section.first())
+            reducedPointList.add(section.last())
+        } else {
+            val firstHalf = reduceSection(section.subList(0, indexMax + 1), epsilon)
+            val secondHalf = reduceSection(section.subList(indexMax, section.size), epsilon)
+            reducedPointList.addAll(firstHalf.subList(0, firstHalf.size - 1))
+            reducedPointList.addAll(secondHalf)
+        }
+        return reducedPointList.toList()
+    }
+
+    /**
+     * Compute the distances between each point the segment (described by start and end).
+     * @param points The points we want to calculate the distance with
+     * @param start The starting point of the segment
+     * @param end The end point of the segment
+     * @return The distance between each point and the segment.
+     */
+    private fun distancesToSegment(points: List<LatLng>, start: LatLng, end: LatLng): List<Float> {
+        // Translate to put the start at the origin
+        var newEnd = LatLng(end.latitude - start.latitude, end.longitude - start.longitude)
+        // Consider the angle in the plane perpendicular to the x axis in spherical coordinates
+        // So the angle is atan2(z, y) where z = sin(latitude), y = cos(latitude)sin(longitude)
+        val anglePlane = atan2(
+            sin(newEnd.latitude * PI / 180),
+            cos(newEnd.latitude * PI / 180) * sin(newEnd.longitude * PI / 180),
+        )
+        // Rotate the end point to be aligned with the equator
+        newEnd = rotateXAxis(newEnd.latitude, newEnd.longitude, -anglePlane)
+
+        var distances = mutableListOf<Float>()
+        for (point in points) {
+            // Translate to put the start at the origin and rotate to put the segment on the equator
+            val newPoint = rotateXAxis(
+                point.latitude - start.latitude,
+                point.longitude - start.longitude,
+                -anglePlane,
+            )
+            // Compute the distance to the segment case by case
+            var results = FloatArray(3)
```

Comment:
`distance` and `result` should be val, not var.

Reply:
Right, the content of the list is modified, not the list themselves.